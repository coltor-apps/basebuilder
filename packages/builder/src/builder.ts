import { type Entity } from "./entity";
import { type Input } from "./input";
import { generateUuid, validateUuid } from "./uuid";

export type BuilderEntities = ReadonlyArray<
  Entity<string, ReadonlyArray<Input<string, unknown>>, unknown>
>;

type ChildrenAllowed<TEntities extends BuilderEntities> = {
  [K in TEntities[number]["name"]]?:
    | ReadonlyArray<TEntities[number]["name"]>
    | true;
};

export type Builder<
  TEntities extends BuilderEntities = BuilderEntities,
  TChildrenAllowed extends ChildrenAllowed<TEntities> = Record<
    string,
    true | ReadonlyArray<string> | undefined
  >,
  TParentRequired extends ReadonlyArray<
    TEntities[number]["name"]
  > = ReadonlyArray<string>,
> = {
  entities: TEntities;
  entityId: {
    generate: () => string;
    validate: (id: string) => void;
  };
  childrenAllowed: TChildrenAllowed;
  parentRequired: TParentRequired;
};

type OptionalBuilderArgs = "childrenAllowed" | "parentRequired";

export function createBuilder<
  const TEntities extends BuilderEntities,
  const TChildrenAllowed extends ChildrenAllowed<TEntities> = Record<
    string,
    never
  >,
  const TParentRequired extends ReadonlyArray<TEntities[number]["name"]> = [],
>(
  options: Omit<
    Builder<TEntities, TChildrenAllowed, TParentRequired>,
    OptionalBuilderArgs | "entityId"
  > &
    Partial<
      Pick<
        Builder<TEntities, TChildrenAllowed, TParentRequired>,
        OptionalBuilderArgs
      >
    > & {
      entityId?: Partial<
        Builder<TEntities, TChildrenAllowed, TParentRequired>["entityId"]
      >;
    },
): Builder<TEntities, TChildrenAllowed, TParentRequired> {
  return {
    ...options,
    entityId: {
      generate: options.entityId?.generate ?? generateUuid,
      validate: options.entityId?.validate ?? validateUuid,
    },
    childrenAllowed: options.childrenAllowed ?? ({} as TChildrenAllowed),
    parentRequired:
      options.parentRequired ??
      ([] as ReadonlyArray<unknown> as TParentRequired),
  };
}

export function getEntityDefinition(
  entityType: string,
  builder: Builder,
): Builder["entities"][number] | undefined {
  return builder.entities.find(
    (builderEntity) => builderEntity.name === entityType,
  );
}

export function ensureEntityIsRegistered(
  entityType: string,
  builder: Builder,
): Builder["entities"][number] {
  const entityDefinition = getEntityDefinition(entityType, builder);

  if (!entityDefinition) {
    throw new Error(`Unkown entity type "${entityType}".`);
  }

  return entityDefinition;
}

export function ensureEntityInputIsRegistered(
  entityType: string,
  inputName: string,
  builder: Builder,
): Input {
  const entityDefinition = ensureEntityIsRegistered(entityType, builder);

  const input = entityDefinition.inputs.find(
    (input) => input.name === inputName,
  );

  if (!input) {
    throw new Error(`Unkown entity input "${inputName}".`);
  }

  return input;
}

export function ensureEntityInputsAreRegistered(
  entityType: string,
  inputNames: Array<string>,
  builder: Builder,
): Array<Input> {
  const inputs = inputNames.map((inputName) =>
    ensureEntityInputIsRegistered(entityType, inputName, builder),
  );

  return inputs;
}

export function isEntityChildAllowed(
  entityType: string,
  childEntityType: string,
  builder: Builder,
): boolean {
  const allowedChildren = builder.childrenAllowed[entityType];

  if (!allowedChildren) {
    return false;
  }

  return allowedChildren === true || allowedChildren.includes(childEntityType);
}

export function isEntityParentRequired(
  entityType: string,
  builder: Builder,
): boolean {
  return builder.parentRequired.includes(entityType);
}

export function ensureEntityChildAllowed(
  entityType: string,
  childEntityType: string,
  builder: Builder,
): void {
  if (!isEntityChildAllowed(entityType, childEntityType, builder)) {
    throw new Error("Child is not allowed.");
  }
}

export function ensureEntityCanLackParent(
  entityType: string,
  builder: Builder,
): void {
  if (isEntityParentRequired(entityType, builder)) {
    throw new Error("A parent is required.");
  }
}
