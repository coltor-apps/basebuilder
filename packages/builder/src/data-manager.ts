import {
  createSubscriptionManager,
  type Subscribe,
} from "./subscription-manager";

export function createDataManager<TData>(initialData: TData): {
  getData: () => TData;
  setData: (data: TData) => TData;
  subscribe: Subscribe<TData>;
} {
  let data: TData = initialData;

  function getData(): TData {
    return data;
  }

  const { notify, subscribe } = createSubscriptionManager<TData>();

  return {
    subscribe,
    getData,
    setData(newData) {
      data = newData;

      notify(data);

      return data;
    },
  };
}
