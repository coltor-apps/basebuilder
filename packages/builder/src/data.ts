import { createSubscriptionManager, type Subscribe } from "./subscription";

export function createDataManager<TData>(data: TData): {
  getData: () => TData;
  setData: (setter: (oldData: TData) => TData) => TData;
  subscribe: Subscribe<TData>;
} {
  let localData: TData = data;

  function getData(): TData {
    return localData;
  }

  const { notify, subscribe } = createSubscriptionManager<TData>();

  return {
    subscribe,
    getData,
    setData(setter) {
      localData = setter(getData());

      notify(localData);

      return localData;
    },
  };
}
