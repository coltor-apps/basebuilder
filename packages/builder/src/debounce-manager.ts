export type DebounceManager<TResult> = {
  handle(
    key: string,
    callback: () => Promise<TResult> | TResult,
    fallback: () => TResult,
  ): Promise<TResult>;
  get(key: string): Date | undefined;
};

function isTimestampCurrent(
  timestamps: Map<string, Date>,
  key: string,
  timestamp: Date,
): boolean {
  return timestamps.get(key)?.getTime() !== timestamp.getTime();
}

export function createDebounceManager<TResult>(): DebounceManager<TResult> {
  const timestamps: Map<string, Date> = new Map();

  return {
    async handle(key, callback, fallback) {
      const currentTimestamp = new Date();

      timestamps.set(key, currentTimestamp);

      try {
        const result = await callback();

        if (isTimestampCurrent(timestamps, key, currentTimestamp)) {
          return fallback();
        }

        return result;
      } catch (error) {
        throw error;
      } finally {
        if (isTimestampCurrent(timestamps, key, currentTimestamp)) {
          timestamps.delete(key);
        }
      }
    },
    get(key) {
      return timestamps.get(key);
    },
  };
}
