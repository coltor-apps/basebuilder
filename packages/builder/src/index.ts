export { createBuilder } from "./builder";
export type { Builder, BuilderEntities } from "./builder";

export { createEntity } from "./entity";
export type { Entity } from "./entity";

export { createInput } from "./input";
export type { Input } from "./input";

export { SchemaValidationError as EntityValidationError, createStore } from "./store";
export type {
  Store,
  StoreData,
  StoreEntity,
  SchemaValidationErrorCode as EntityValidationErrorCause,
} from "./store";

export { getEntitiesNamesExcept } from "./utils";
