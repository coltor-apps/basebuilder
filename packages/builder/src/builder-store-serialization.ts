import { type Builder } from "./builder";
import {
  type BuilderStoreData,
  type BuilderStoreEntity,
  type SerializedBuilderStoreData,
} from "./builder-store";

export function serializeBuilderStoreData<TBuilder extends Builder>(
  data: BuilderStoreData<TBuilder>,
): SerializedBuilderStoreData<TBuilder> {
  const newEntities: SerializedBuilderStoreData<TBuilder>["schema"]["entities"] =
    {};

  for (const [id, entity] of data.schema.entities) {
    const { children, ...entityData } = entity;

    newEntities[id] = {
      ...entityData,
      ...(children ? { children: Array.from(children) } : {}),
      inputs:
        entityData.inputs as unknown as SerializedBuilderStoreData<TBuilder>["schema"]["entities"][string]["inputs"],
    };
  }

  return {
    schema: {
      root: Array.from(data.schema.root),
      entities: newEntities,
    },
    entitiesInputsErrors: Object.fromEntries(data.entitiesInputsErrors),
  };
}

export function deserializeBuilderStoreData<TBuilder extends Builder>(
  data: SerializedBuilderStoreData<TBuilder>,
): BuilderStoreData<TBuilder> {
  return {
    schema: {
      entities: new Map(
        Object.entries(data.schema.entities).map(([id, entity]) => [
          id,
          {
            ...entity,
            ...(entity.children ? { children: new Set(entity.children) } : {}),
            inputs:
              entity.inputs as unknown as BuilderStoreEntity<TBuilder>["inputs"],
          },
        ]),
      ),
      root: new Set(data.schema.root),
    },
    entitiesInputsErrors: new Map(Object.entries(data.entitiesInputsErrors)),
  };
}
