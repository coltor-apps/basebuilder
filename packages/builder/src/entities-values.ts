import { ensureEntityIsRegistered, type Builder } from "./builder";
import { ensureEntityExists, type Schema } from "./schema";

export type EntityValue<TBuilder extends Builder = Builder> =
  | Awaited<ReturnType<TBuilder["entities"][number]["validate"]>>
  | undefined;

export type EntitiesValues<TBuilder extends Builder = Builder> = Record<
  string,
  EntityValue<TBuilder>
>;

export type EntitiesErrors = Record<string, unknown>;

export type EntitiesValuesValidationResult<TBuilder extends Builder> =
  | { data: EntitiesValues<TBuilder>; success: true }
  | { entitiesErrors: EntitiesErrors; success: false };

export async function validateEntityValue<TBuilder extends Builder>(
  entityId: string,
  entitiesValues: EntitiesValues<TBuilder>,
  builder: TBuilder,
  schema: Schema<TBuilder>,
): Promise<
  | { value: EntityValue<TBuilder>; success: true }
  | { error: unknown; success: false }
> {
  const entity = ensureEntityExists(entityId, schema.entities);

  const entityDefinition = ensureEntityIsRegistered(entity.type, builder);

  try {
    const value = (await entityDefinition.validate(entitiesValues[entityId], {
      entity,
      entitiesValues,
    })) as EntityValue<TBuilder>;

    return { success: true, value };
  } catch (error) {
    return { success: false, error };
  }
}

function getEligibleEntitiesIdsForValidationFromEntity<
  TBuilder extends Builder,
>(
  entityId: string,
  entitiesValues: EntitiesValues<TBuilder>,
  schema: Schema<TBuilder>,
  builder: TBuilder,
): string[] {
  const entity = ensureEntityExists(entityId, schema.entities);

  const entityDefinition = ensureEntityIsRegistered(entity.type, builder);

  const shouldBeProcessed = entityDefinition.shouldBeProcessed({
    entity,
    entitiesValues,
  });

  let eligibleEntities: string[] = [];

  if (entityDefinition.valueAllowed && shouldBeProcessed) {
    eligibleEntities.push(entityId);
  }

  if (entity.children && shouldBeProcessed) {
    for (const childId of entity.children) {
      eligibleEntities = eligibleEntities.concat(
        getEligibleEntitiesIdsForValidationFromEntity(
          childId,
          entitiesValues,
          schema,
          builder,
        ),
      );
    }
  }

  return eligibleEntities;
}

export function getEligibleEntitiesIdsForValidation<TBuilder extends Builder>(
  entitiesValues: EntitiesValues<TBuilder>,
  builder: TBuilder,
  schema: Schema<TBuilder>,
): string[] {
  let eligibleEntities: string[] = [];

  for (const rootEntityId of schema.root) {
    eligibleEntities = eligibleEntities.concat(
      getEligibleEntitiesIdsForValidationFromEntity(
        rootEntityId,
        entitiesValues,
        schema,
        builder,
      ),
    );
  }

  return eligibleEntities;
}

export async function validateEntitiesValues<TBuilder extends Builder>(
  entitiesValues: unknown,
  builder: TBuilder,
  schema: Schema<TBuilder>,
): Promise<EntitiesValuesValidationResult<TBuilder>> {
  const computedEntitiesValues =
    typeof entitiesValues !== "object" ||
    Array.isArray(entitiesValues) ||
    entitiesValues === null
      ? {}
      : (entitiesValues as EntitiesValues<TBuilder>);

  const eligibleEntitiesIdsForValidation = getEligibleEntitiesIdsForValidation(
    computedEntitiesValues,
    builder,
    schema,
  );

  const entitiesErrors: EntitiesErrors = {};

  const newEntitiesValues: EntitiesValues<TBuilder> = {
    ...computedEntitiesValues,
  };

  for (const entityId in newEntitiesValues) {
    if (!eligibleEntitiesIdsForValidation.includes(entityId)) {
      delete newEntitiesValues[entityId];
    }
  }

  for (const entityId of eligibleEntitiesIdsForValidation) {
    const validationResult = await validateEntityValue(
      entityId,
      newEntitiesValues,
      builder,
      schema,
    );

    if (!validationResult.success) {
      entitiesErrors[entityId] = validationResult.error;
    } else {
      newEntitiesValues[entityId] = validationResult.value;
    }
  }

  if (Object.keys(entitiesErrors).length) {
    return {
      success: false,
      entitiesErrors,
    };
  }

  return { success: true, data: newEntitiesValues };
}
