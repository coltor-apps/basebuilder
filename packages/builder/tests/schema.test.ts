import { describe, expect, it } from "vitest";
import { z } from "zod";

import { createBuilder, createEntity, createInput } from "../src";
import {
  SchemaValidationError,
  schemaValidationErrorCodes,
  validateSchema,
  validateSchemaIntegrity,
  type Schema,
  type SchemaValidationErrorCause,
} from "../src/schema";

const invalidSchemasCases: Array<{
  schema: unknown;
  errorCause: SchemaValidationErrorCause;
}> = [
  {
    schema: {
      entities: {},
    },
    errorCause: {
      code: schemaValidationErrorCodes.InvalidRootFormat,
    },
  },
  {
    schema: {
      entities: {},
      root: {},
    },
    errorCause: {
      code: schemaValidationErrorCodes.InvalidRootFormat,
      root: {},
    },
  },
  {
    schema: {
      entities: {},
      root: null,
    },
    errorCause: {
      code: schemaValidationErrorCodes.InvalidRootFormat,
      root: null,
    },
  },
  {
    schema: {
      entities: {
        "c1ab14a4-41db-4531-9a58-4825a9ef6d26": {
          type: "text",
          inputs: {},
        },
      },
      root: [
        "c1ab14a4-41db-4531-9a58-4825a9ef6d26",
        "c1ab14a4-41db-4531-9a58-4825a9ef6d26",
      ],
    },
    errorCause: {
      code: schemaValidationErrorCodes.DuplicateRootId,
      entityId: "c1ab14a4-41db-4531-9a58-4825a9ef6d26",
    },
  },
  {
    schema: {
      entities: {
        entities: {
          "c1ab14a4-41db-4531-9a58-4825a9ef6d26": {
            type: "text",
            inputs: {},
          },
        },
      },
      root: [],
    },
    errorCause: {
      code: schemaValidationErrorCodes.EmptyRoot,
    },
  },
  {
    schema: {
      entities: {},
      root: ["c1ab14a4-41db-4531-9a58-4825a9ef6d26"],
    },
    errorCause: {
      code: schemaValidationErrorCodes.NonexistentEntityId,
      entityId: "c1ab14a4-41db-4531-9a58-4825a9ef6d26",
    },
  },
  {
    schema: {
      entities: [],
      root: [],
    },
    errorCause: {
      code: schemaValidationErrorCodes.InvalidEntitiesFormat,
      entities: [],
    },
  },
  {
    schema: {
      entities: null,
      root: [],
    },
    errorCause: {
      code: schemaValidationErrorCodes.InvalidEntitiesFormat,
      entities: null,
    },
  },
  {
    schema: {
      entities: {
        "c1ab14a4-41db-4531-9a58-4825a9ef6d26": {
          inputs: {},
        },
      },
      root: ["c1ab14a4-41db-4531-9a58-4825a9ef6d26"],
    },
    errorCause: {
      code: schemaValidationErrorCodes.MissingEntityType,
      entityId: "c1ab14a4-41db-4531-9a58-4825a9ef6d26",
    },
  },
  {
    schema: {
      entities: {
        "c1ab14a4-41db-4531-9a58-4825a9ef6d26": {
          type: "invalid",
          inputs: {},
        },
      },
      root: ["c1ab14a4-41db-4531-9a58-4825a9ef6d26"],
    },
    errorCause: {
      code: schemaValidationErrorCodes.UnknownEntityType,
      entityType: "invalid",
      entityId: "c1ab14a4-41db-4531-9a58-4825a9ef6d26",
    },
  },
  {
    schema: {
      entities: {
        "c1ab14a4-41db-4531-9a58-4825a9ef6d26": {
          type: "text",
        },
      },
      root: ["c1ab14a4-41db-4531-9a58-4825a9ef6d26"],
    },
    errorCause: {
      code: schemaValidationErrorCodes.MissingEntityInputs,
      entityId: "c1ab14a4-41db-4531-9a58-4825a9ef6d26",
    },
  },
  {
    schema: {
      entities: {
        "c1ab14a4-41db-4531-9a58-4825a9ef6d26": {
          type: "text",
          inputs: [],
        },
      },
      root: ["c1ab14a4-41db-4531-9a58-4825a9ef6d26"],
    },
    errorCause: {
      code: schemaValidationErrorCodes.InvalidEntityInputsFormat,
      entityId: "c1ab14a4-41db-4531-9a58-4825a9ef6d26",
      entityInputs: [],
    },
  },
  {
    schema: {
      entities: {
        "c1ab14a4-41db-4531-9a58-4825a9ef6d26": {
          type: "text",
          inputs: {
            invalid: "test",
          },
        },
      },
      root: ["c1ab14a4-41db-4531-9a58-4825a9ef6d26"],
    },
    errorCause: {
      code: schemaValidationErrorCodes.UnknownEntityInputType,
      entityId: "c1ab14a4-41db-4531-9a58-4825a9ef6d26",
      inputName: "invalid",
    },
  },
  {
    schema: {
      entities: {
        "c1ab14a4-41db-4531-9a58-4825a9ef6d26": {
          type: "text",
          parentId: "6e0035c3-0d4c-445f-a42b-2d971225447c",
          inputs: {},
        },
      },
      root: ["c1ab14a4-41db-4531-9a58-4825a9ef6d26"],
    },
    errorCause: {
      code: schemaValidationErrorCodes.NonexistentEntityParent,
      entityId: "c1ab14a4-41db-4531-9a58-4825a9ef6d26",
      entityParentId: "6e0035c3-0d4c-445f-a42b-2d971225447c",
    },
  },
  {
    schema: {
      entities: {
        "c1ab14a4-41db-4531-9a58-4825a9ef6d26": {
          type: "text",
          parentId: "c1ab14a4-41db-4531-9a58-4825a9ef6d26",
          inputs: {},
        },
      },
      root: ["c1ab14a4-41db-4531-9a58-4825a9ef6d26"],
    },
    errorCause: {
      code: schemaValidationErrorCodes.SelfEntityReference,
      entityId: "c1ab14a4-41db-4531-9a58-4825a9ef6d26",
    },
  },
  {
    schema: {
      entities: {
        "c1ab14a4-41db-4531-9a58-4825a9ef6d26": {
          type: "text",
          inputs: {},
          children: "invalid",
        },
      },
      root: ["c1ab14a4-41db-4531-9a58-4825a9ef6d26"],
    },
    errorCause: {
      code: schemaValidationErrorCodes.InvalidChildrenFormat,
      entityId: "c1ab14a4-41db-4531-9a58-4825a9ef6d26",
    },
  },
  {
    schema: {
      entities: {
        "c1ab14a4-41db-4531-9a58-4825a9ef6d26": {
          type: "text",
          inputs: {},
          children: ["6e0035c3-0d4c-445f-a42b-2d971225447c"],
        },
        "6e0035c3-0d4c-445f-a42b-2d971225447c": {
          type: "text",
          inputs: {},
          parentId: "c1ab14a4-41db-4531-9a58-4825a9ef6d26",
        },
      },
      root: ["c1ab14a4-41db-4531-9a58-4825a9ef6d26"],
    },
    errorCause: {
      code: schemaValidationErrorCodes.ChildNotAllowed,
      entityId: "c1ab14a4-41db-4531-9a58-4825a9ef6d26",
      childId: "6e0035c3-0d4c-445f-a42b-2d971225447c",
    },
  },
  {
    schema: {
      entities: {
        "6e0035c3-0d4c-445f-a42b-2d971225447c": {
          type: "section",
          inputs: {},
          parentId: "a1109529-46c6-4290-885b-bb0aca7a92a1",
        },
        "a1109529-46c6-4290-885b-bb0aca7a92a1": {
          type: "section",
          inputs: {},
          children: ["6e0035c3-0d4c-445f-a42b-2d971225447c"],
        },
      },
      root: [
        "6e0035c3-0d4c-445f-a42b-2d971225447c",
        "a1109529-46c6-4290-885b-bb0aca7a92a1",
      ],
    },
    errorCause: {
      code: schemaValidationErrorCodes.RootEntityWithParent,
      entityId: "6e0035c3-0d4c-445f-a42b-2d971225447c",
    },
  },
  {
    schema: {
      entities: {
        "c1ab14a4-41db-4531-9a58-4825a9ef6d26": {
          type: "text",
          inputs: {},
        },
        "6e0035c3-0d4c-445f-a42b-2d971225447c": {
          type: "section",
          inputs: {},
          children: ["c1ab14a4-41db-4531-9a58-4825a9ef6d26"],
        },
      },
      root: [
        "6e0035c3-0d4c-445f-a42b-2d971225447c",
        "c1ab14a4-41db-4531-9a58-4825a9ef6d26",
      ],
    },
    errorCause: {
      code: schemaValidationErrorCodes.EntityChildrenMismatch,
      entityId: "6e0035c3-0d4c-445f-a42b-2d971225447c",
      childId: "c1ab14a4-41db-4531-9a58-4825a9ef6d26",
    },
  },
  {
    schema: {
      entities: {
        "c1ab14a4-41db-4531-9a58-4825a9ef6d26": {
          type: "text",
          inputs: {},
          parentId: "6e0035c3-0d4c-445f-a42b-2d971225447c",
        },
        "6e0035c3-0d4c-445f-a42b-2d971225447c": {
          type: "section",
          inputs: {},
          children: [
            "c1ab14a4-41db-4531-9a58-4825a9ef6d26",
            "c1ab14a4-41db-4531-9a58-4825a9ef6d26",
          ],
        },
      },
      root: ["6e0035c3-0d4c-445f-a42b-2d971225447c"],
    },
    errorCause: {
      code: schemaValidationErrorCodes.DuplicateChildId,
      entityId: "6e0035c3-0d4c-445f-a42b-2d971225447c",
    },
  },
  {
    schema: {
      entities: {
        "c1ab14a4-41db-4531-9a58-4825a9ef6d26": {
          type: "text",
          inputs: {},
          parentId: "6e0035c3-0d4c-445f-a42b-2d971225447c",
        },
        "6e0035c3-0d4c-445f-a42b-2d971225447c": {
          type: "section",
          inputs: {},
          children: ["a1109529-46c6-4290-885b-bb0aca7a92a1"],
        },
        "a1109529-46c6-4290-885b-bb0aca7a92a1": {
          type: "section",
          inputs: {},
        },
      },
      root: ["6e0035c3-0d4c-445f-a42b-2d971225447c"],
    },
    errorCause: {
      code: schemaValidationErrorCodes.EntityParentMismatch,
      entityId: "c1ab14a4-41db-4531-9a58-4825a9ef6d26",
      parentId: "6e0035c3-0d4c-445f-a42b-2d971225447c",
    },
  },
  {
    schema: {
      entities: {
        "a1109529-46c6-4290-885b-bb0aca7a92a1": {
          type: "select",
          inputs: {},
        },
      },
      root: ["a1109529-46c6-4290-885b-bb0aca7a92a1"],
    },
    errorCause: {
      code: schemaValidationErrorCodes.ParentRequired,
      entityId: "a1109529-46c6-4290-885b-bb0aca7a92a1",
    },
  },
];

describe("schema integrity validation", () => {
  it("throws for invalid schemas", () => {
    const builder = createBuilder({
      entities: [
        createEntity({
          name: "text",
          inputs: [
            createInput({
              name: "label",
              validate(value) {
                return value;
              },
            }),
          ],
        }),
        createEntity({
          name: "section",
        }),
        createEntity({
          name: "select",
        }),
      ],
      childrenAllowed: {
        section: ["text", "section"],
      },
      parentRequired: ["select"],
    });

    for (const item of invalidSchemasCases) {
      const result = validateSchemaIntegrity(item.schema as Schema, {
        builder,
      });

      expect(result).toEqual({
        success: false,
        error: new SchemaValidationError(item.errorCause),
      });
    }
  });

  it("returns the validated schema", () => {
    const builder = createBuilder({
      entities: [
        createEntity({
          name: "text",
          inputs: [
            createInput({
              name: "label",
              validate(value) {
                return z.string().parse(value) + "should be appended";
              },
            }),
          ],
        }),
      ],
    });

    expect(
      validateSchemaIntegrity(
        {
          entities: {
            "c1ab14a4-41db-4531-9a58-4825a9ef6d26": {
              type: "text",
              // @ts-expect-error Intentional wrong data type.
              inputs: {},
            },
          },
          root: ["c1ab14a4-41db-4531-9a58-4825a9ef6d26"],
        },
        {
          builder,
        },
      ),
    ).toMatchSnapshot();
  });

  it("throws for invalid parent id", () => {
    const builder = createBuilder({
      entities: [
        createEntity({
          name: "text",
        }),
      ],
    });

    expect(() =>
      validateSchemaIntegrity(
        {
          entities: {
            "c1ab14a4-41db-4531-9a58-4825a9ef6d26": {
              type: "text",
              inputs: {},
              // @ts-expect-error Intentional wrong data type.
              parentId: 1,
            },
          },
          root: ["c1ab14a4-41db-4531-9a58-4825a9ef6d26"],
        },
        {
          builder,
        },
      ),
    ).toThrowErrorMatchingSnapshot();
  });

  it("throws for invalid children ids", () => {
    const builder = createBuilder({
      entities: [
        createEntity({
          name: "text",
        }),
      ],
      childrenAllowed: {
        text: true,
      },
    });

    expect(() =>
      validateSchemaIntegrity(
        {
          entities: {
            "c1ab14a4-41db-4531-9a58-4825a9ef6d26": {
              type: "text",
              inputs: {},
              // @ts-expect-error Intentional wrong data type.
              children: [1],
            },
          },
          root: ["c1ab14a4-41db-4531-9a58-4825a9ef6d26"],
        },
        {
          builder,
        },
      ),
    ).toThrowErrorMatchingSnapshot();
  });

  it("throws for invalid root ids", () => {
    const builder = createBuilder({
      entities: [
        createEntity({
          name: "text",
        }),
      ],
    });

    expect(() =>
      validateSchemaIntegrity(
        {
          entities: {
            "c1ab14a4-41db-4531-9a58-4825a9ef6d26": {
              type: "text",
              inputs: {},
            },
          },
          // @ts-expect-error Intentional wrong data type.
          root: [1, "c1ab14a4-41db-4531-9a58-4825a9ef6d26"],
        },
        {
          builder,
        },
      ),
    ).toThrowErrorMatchingSnapshot();
  });

  it("returns clean data for valid schemas", () => {
    const textEntity = createEntity({
      name: "text",
      inputs: [
        createInput({
          name: "label",
          validate(value) {
            return value;
          },
        }),
      ],
    });

    const sectionEntity = createEntity({
      name: "section",
    });

    const builder = createBuilder({
      entities: [textEntity, sectionEntity],
      childrenAllowed: {
        section: true,
      },
    });

    const schema: Schema = {
      entities: {
        "c1ab14a4-41db-4531-9a58-4825a9ef6d26": {
          type: "text",
          inputs: {
            label: "test",
          },
          parentId: "6e0035c3-0d4c-445f-a42b-2d971225447c",
          // @ts-expect-error Intentionally redundant property.
          dirty: "should be removed after validation",
        },
        "6e0035c3-0d4c-445f-a42b-2d971225447c": {
          type: "section",
          inputs: {},
          children: ["c1ab14a4-41db-4531-9a58-4825a9ef6d26"],
        },
      },
      root: ["6e0035c3-0d4c-445f-a42b-2d971225447c"],
    };

    expect(validateSchemaIntegrity(schema, { builder })).toMatchSnapshot();
  });
});

describe("schema validation", () => {
  it("throws for invalid schemas", async () => {
    const builder = createBuilder({
      entities: [
        createEntity({
          name: "text",
          inputs: [
            createInput({
              name: "label",
              validate(value) {
                return value;
              },
            }),
          ],
        }),
        createEntity({
          name: "section",
        }),
        createEntity({
          name: "select",
        }),
      ],
      childrenAllowed: {
        section: ["text", "section"],
      },
      parentRequired: ["select"],
    });

    for (const item of invalidSchemasCases) {
      const result = await validateSchema(item.schema as Schema, { builder });

      expect(result).toEqual({
        success: false,
        error: new SchemaValidationError(item.errorCause),
      });
    }
  });

  it("validates inputs with their validators", async () => {
    const builder = createBuilder({
      entities: [
        createEntity({
          name: "text",
          inputs: [
            createInput({
              name: "label",
              validate(value) {
                return z.string().parse(value);
              },
            }),
          ],
        }),
      ],
    });

    const result = await validateSchema(
      {
        entities: {
          "c1ab14a4-41db-4531-9a58-4825a9ef6d26": {
            type: "text",
            inputs: {
              // @ts-expect-error Intentional wrong data type
              label: 1,
            },
          },
          "4b9ed44b-0e4d-41e9-ad73-1ee70e8fefcb": {
            type: "text",
            inputs: {
              // @ts-expect-error Intentional wrong data type
              label: 1,
            },
          },
        },
        root: [
          "c1ab14a4-41db-4531-9a58-4825a9ef6d26",
          "4b9ed44b-0e4d-41e9-ad73-1ee70e8fefcb",
        ],
      },
      { builder },
    );

    expect(result).toMatchSnapshot();

    expect(
      ((result as Record<string, unknown>).error as Record<string, unknown>)
        .cause,
    ).toMatchSnapshot();
  });
});
