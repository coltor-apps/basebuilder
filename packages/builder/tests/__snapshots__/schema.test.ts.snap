// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`schema integrity validation > returns clean data for valid schemas 1`] = `
{
  "data": {
    "entities": {
      "6e0035c3-0d4c-445f-a42b-2d971225447c": {
        "children": [
          "c1ab14a4-41db-4531-9a58-4825a9ef6d26",
        ],
        "inputs": {},
        "type": "section",
      },
      "c1ab14a4-41db-4531-9a58-4825a9ef6d26": {
        "inputs": {
          "label": "test",
        },
        "parentId": "6e0035c3-0d4c-445f-a42b-2d971225447c",
        "type": "text",
      },
    },
    "root": [
      "6e0035c3-0d4c-445f-a42b-2d971225447c",
    ],
  },
  "success": true,
}
`;

exports[`schema integrity validation > returns the validated schema 1`] = `
{
  "data": {
    "entities": {
      "c1ab14a4-41db-4531-9a58-4825a9ef6d26": {
        "inputs": {},
        "type": "text",
      },
    },
    "root": [
      "c1ab14a4-41db-4531-9a58-4825a9ef6d26",
    ],
  },
  "success": true,
}
`;

exports[`schema integrity validation > throws for invalid children ids 1`] = `"The entity id '1' is invalid."`;

exports[`schema integrity validation > throws for invalid parent id 1`] = `"The entity id '1' is invalid."`;

exports[`schema integrity validation > throws for invalid root ids 1`] = `"The entity id '1' is invalid."`;

exports[`schema validation > validates inputs with their validators 1`] = `
{
  "reason": {
    "code": "InvalidEntitiesInputs",
    "payload": {
      "entitiesInputsErrors": {
        "4b9ed44b-0e4d-41e9-ad73-1ee70e8fefcb": {
          "label": [ZodError: [
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [],
    "message": "Expected string, received number"
  }
]],
        },
        "c1ab14a4-41db-4531-9a58-4825a9ef6d26": {
          "label": [ZodError: [
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [],
    "message": "Expected string, received number"
  }
]],
        },
      },
    },
  },
  "success": false,
}
`;

exports[`schema validation > validates inputs with their validators 2`] = `
{
  "entitiesInputsErrors": {
    "4b9ed44b-0e4d-41e9-ad73-1ee70e8fefcb": {
      "label": [ZodError: [
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [],
    "message": "Expected string, received number"
  }
]],
    },
    "c1ab14a4-41db-4531-9a58-4825a9ef6d26": {
      "label": [ZodError: [
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [],
    "message": "Expected string, received number"
  }
]],
    },
  },
}
`;
